python/ray/data/impl/push_based_shuffle.py

```diff
@@ -0,0 +1,388 @@
+        clear_input_blocks: bool,
+        *,
+        map_ray_remote_args: Optional[Dict[str, Any]] = None,
+        reduce_ray_remote_args: Optional[Dict[str, Any]] = None,
+        merge_factor: int = 2,
+    ) -> Tuple[BlockList, Dict[str, List[BlockMetadata]]]:
+        logger.info("Using experimental push-based shuffle.")
+        # TODO(swang): For jobs whose reduce work is heavier than the map work,
+        # we should support fractional merge factors.
+        # TODO(swang): For large jobs, we should try to choose the merge factor
+        # automatically, e.g., by running one test round of map and merge tasks
+        # and comparing their run times.
+        # TODO(swang): Add option to automatically reduce write amplification
+        # during map-merge stage, by limiting how many partitions can be
+        # processed concurrently.
+        input_blocks_list = input_blocks.get_blocks()
+        input_num_blocks = len(input_blocks_list)
+
+        if map_ray_remote_args is None:
+            map_ray_remote_args = {}
+        if reduce_ray_remote_args is None:
+            reduce_ray_remote_args = {}
+        # The placement strategy for reduce tasks is overwritten to colocate
+        # them with their inputs from the merge stage, so remove any
+        # pre-specified scheduling strategy here.
+        reduce_ray_remote_args.pop("scheduling_strategy", None)
+
+        map_fn = self._map_partition
+        merge_fn = self._merge
+
+        def map_partition(*args, **kwargs):
+            return map_fn(self.map, *args, **kwargs)
+
+        def merge(*args, **kwargs):
+            return merge_fn(self.reduce, *args, **kwargs)
+
+        shuffle_map = cached_remote_fn(map_partition)
+        shuffle_merge = cached_remote_fn(merge)
+        shuffle_reduce = cached_remote_fn(self.reduce)
+
+        # Compute all constants used for task scheduling.
+        cpu_map = PushBasedShufflePlan._get_cluster_cpu_map()
+        num_cpus_total = sum(v for v in cpu_map.values())
+        task_parallelism = min(num_cpus_total, input_num_blocks)
+        # Total number of merge tasks.
+        num_merge_tasks = math.ceil(input_num_blocks / merge_factor)
+        # Number of map-merge rounds.
+        num_rounds = math.ceil((input_num_blocks + num_merge_tasks) / task_parallelism)
+        # M: Number of map tasks in one map-merge round.
+        num_map_tasks_per_round = math.ceil(input_num_blocks / num_rounds)
+        # N: Number of merge tasks in one map-merge round. Each map_partition
+        # task will send one output to each of these merge tasks.
+        num_merge_tasks_per_round = math.ceil(num_merge_tasks / num_rounds)
+        # Scheduling args for assigning merge tasks to nodes. We use node-affinity
+        # scheduling here to colocate merge tasks that output to the same
+        # reducer.
+        merge_task_args = self._compute_merge_task_options(
+            num_merge_tasks_per_round, merge_factor, cpu_map
+        )
```

Comment:
Could we unit test the above topology configuration logic? Concretely, we could split it out into helper method that returns a struct/dict, and we can unit test the topology configured for various cluster scenarios. This would also help with splitting the code up for readability.

Reply:
+1